python/pyspark/sql/pandas/group_ops.py

```diff
@@ -501,7 +502,43 @@ def transformWithStateUDF(
                 )
 
             statefulProcessorApiClient.set_implicit_key(key)
-            result = statefulProcessor.handleInputRows(key, inputRows)
+
+            batch_timestamp = statefulProcessorApiClient.get_batch_timestamp()
+            watermark_timestamp = statefulProcessorApiClient.get_watermark_timestamp()
+            # process with invalid expiry timer info and emit data rows
+            data_iter = statefulProcessor.handleInputRows(
+                key, inputRows, TimerValues(batch_timestamp, watermark_timestamp), ExpiredTimerInfo(False))
+            statefulProcessorApiClient.set_handle_state(
+                StatefulProcessorHandleState.DATA_PROCESSED
+            )
+
+            if timeMode == "processingtime":
+                expiry_list_iter = statefulProcessorApiClient.get_expiry_timers_iterator(batch_timestamp)
+            elif timeMode == "eventtime":
+                expiry_list_iter = statefulProcessorApiClient.get_expiry_timers_iterator(watermark_timestamp)
+            else:
+                expiry_list_iter = []
+
+            result_iter_list = []
+            # process with valid expiry time info and with empty input rows,
+            # only timer related rows will be emitted
+            for expiry_list in expiry_list_iter:
+                for key_obj, expiry_timestamp in expiry_list:
+                    if timeMode == "processingtime" and expiry_timestamp < batch_timestamp:
+                        result_iter_list.append(statefulProcessor.handleInputRows(
+                            (key_obj,), iter([]),
+                            TimerValues(batch_timestamp, watermark_timestamp),
+                            ExpiredTimerInfo(True, expiry_timestamp)))
+                    elif timeMode == "eventtime" and expiry_timestamp < watermark_timestamp:
+                        result_iter_list.append(statefulProcessor.handleInputRows(
+                            (key_obj,), iter([]),
+                            TimerValues(batch_timestamp, watermark_timestamp),
+                            ExpiredTimerInfo(True, expiry_timestamp)))
+
+            # TODO(SPARK-49603) set the handle state in the lazily initialized iterator
+
+            result_iter_list.insert(0, data_iter)
```

Comment:
Nit: can we just append `data_iter` to the empty list before adding timer iters to it?